README.md
<h1 align="center">
  🚲 <br>
  Velo-action
</h1>

<p align="center">
  A GitHub Action to create releases and trigger deployments
</p>

This GitHub actions is part of the Velo deploy system at Oda.

This repo is public since GitHub actions does not yet
[support actions in private repos](https://github.com/github/roadmap/issues/74).

## Local execution

[The Makefile](../Makefile) covers the most common commands you need. For testing,
you should check out the [example-deploy-project](https://github.com/kolonialno/example-deploy-project/) to have a working deployable application.

 1. Setup dependencies and virtual env: `make install`
 2. All configuration os done via [development variables](../env.dev-vars). It
    should be easy to integrate this env-file into your IDE as well

Now you can execute velo-action using `make run`.

In order to use the containerized build, use `make run_docker`.

## Testing

This can be used to test a pre-release or a PR draft.

1. Create a PR.
2. You may need to update the `action.yml`. The docker image should have the corresponding tag as if you run `make version`. Update `action.yml` with this.
3. Head over to your repository where you want to test `velo-action` in the workflow. Copy the latest commit hash and add it to `<commit>`.

    ```yaml
   - name: Use Velo-action
    uses: kolonialno/velo-action@<commit>
    with:
      service_account_key: ${{ secrets.VELO_ACTION_GSA_KEY_PROD_PUBLIC }}
      ...
    ```

## Create a release

A new draft [Github release of Velo-action](https://github.com/kolonialno/velo-action/releases) is created on every merge to main.

Manually convert the release from a draft to a release to [release it](https://github.com/kolonialno/velo-action/releases).

1. Determine the version of the current release by running `make verison`.
   Bump version if necessary according [SevVer](https://semver.org/) by adding this to the commit message (the PR commits are squashed)

   - `+semver: major`
   - `+semver: minor`

In short:

Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes
MINOR version when you add functionality in a backwards compatible manner
PATCH version when you make backwards compatible bug fixes

Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

1. Update the `changelog.md` with the changes for this release and the upcoming verison.

*IMPORTANT!* A new release will trigger Dependabot. Dependabot will create a PR in all repos that use velo-action.  


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2022-08-23

- Require Google service account as JSON to authenticate with Auth Proxy

## [1.0.17] - 2022-05-11

- Improve workflow and release with comments in PRs.
- Add test verifying version generation when no inputs is provided.
- Do not init tracer when `service_account_key` is not provided. This use case is valid when using `velo-action` to only generate version.

## [1.0.13] - 2022-05-09

Update dependencies.

## [1.0.2] - 2022-04-26

Fix bug preventing the action to output the generated version if `service_account_key` input is not set.

## [1.0.1] - 2022-04-25

### Changed

Fix bug preventing deploys of existing releases in Octopus Deploy. If a release exist, do not recreate it, but do perform the deploys as specified in `deploy_to_environments`.

## [1.0.0] - 2022-04-20

### Added

Breaking changes:

- Input: `project` removed.

  Velo-action will now read the Octopus Project from the AppSpec `app.yml` file as described in the [Velo docs](https://centro.prod.nube.tech/docs/default/component/velo/app-spec/#project).

  Example:

  ```yml
  # .deploy/app.yml
  ...
  project: <project_name>
  vars:
    ...
  ```

- Reverted back to using [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### Changed

- Reformat the releasenotes in Octopus Deploy to use readable HTML formating with links.
- Generate draft Github release of Velo-action on merge to main.

## [0.4.0] - 2022-02-01

### Added

- New input `wait_for_success_seconds` to specify that the action waits until Octopus Deploy is finished.
  Deprecates the input `wait_for_deployment`

### Changed

- Action input `octopus_cli_api_key_secret` is renamed to `octopus_api_key_secret`
- Action input `octopus_cli_server_secret` is renamed to `octopus_server_secret`
- Interact with Octopus Deploy using the API. The Octo CLI is no longer part of the container.

## [0.3.0] - 2021-12-22

### Removed

- Remove generation of semantic version numbers (SemVer). Default is now the shortened git
  hash (`git rev-parse --short HEAD`). Removes although the gitversion dependency. Please see
  [this example](https://github.com/kolonialno/velo/blob/c3d5ddff650fd97357b72ef178d93e5519eb5efa/.github/workflows/ci.yml#L71-L114)
  if you still want to auto-generate the SemVer.

  NOTE: The length of the version string is dynamic. It can be longer if it is not unique.

## [0.2.79] - 2021-12-17

### Changed

- Use `subprocess.run()` when calling the octo CLI

## [0.2.21] - 2021-09-23

## Changed

- Moved image to the new [public Artifact Registry](https://console.cloud.google.com/artifacts/docker/nube-artifacts-prod/europe/nube-container-images-public?project=nube-artifacts-prod) in Google Cloud Platform. Will deprecate the Dockerhub image when there have been no pull for a month. This change require no changes by the user.
- Use python slim image, reduce image size from 1.8GB to 1.14GB.

## [0.2.14] - 2021-08-10

### Added

- Add argument `wait_for_deployment`. This will cause the action to wait until Octopus Deploy finished the deployment. The action will fail if the deploy failes.
- Add argument `progress`. This will show progress of the deployment in Octopus Deploy. In other words a more verbose logs from Octopus Deploy

### Changed

- Refactored the argument parser

## [0.2.6] - 2021-06-30

### Added

- Initial release. Velo-action allows Github Action Workflows to generate a semantic versioning using [GitVersion](https://gitversion.net/), create an imutable relese using Velo and trigger a deploy of that relese in Octopus Deploy.


velo_action/tests/test_tracing.py
import json
import os
import re
import subprocess
from datetime import datetime, timedelta
from pathlib import Path
from unittest.mock import MagicMock, patch

import pytest
from opentelemetry import trace
from opentelemetry.sdk.resources import SERVICE_NAME, Resource  # type: ignore
from opentelemetry.sdk.trace import TracerProvider  # type: ignore
from opentelemetry.sdk.trace.export import (  # type: ignore
    BatchSpanProcessor,
    ConsoleSpanExporter,
)

from velo_action.settings import GithubSettings
from velo_action.tracing_helpers import construct_github_action_trace


def gh_token():
    if token := os.environ.get("GITHUB_TOKEN"):
        return token
    try:
        result = subprocess.run(
            ["gh", "auth", "status", "-t"], capture_output=True, text=True, check=True
        )
        return re.search(r"Token: (.*)\n", result.stderr).group(1)
    except Exception:  # pylint: disable=broad-except
        return None


has_token = pytest.mark.skipif(not bool(gh_token()), reason="No Github token found")


class SpanList:
    def __init__(self):
        self.found_string_spans = []

    def write(self, span):
        span = json.loads(span)

        del span["resource"]["build.repository"]
        del span["resource"]["build.actor"]
        del span["context"]["trace_id"]
        del span["context"]["span_id"]
        del span["parent_id"]

        # Disregard dynamic start and end times
        utc_local_diff = datetime.now() - datetime.utcnow() + timedelta(seconds=1)
        if (
            datetime.now() - datetime.fromisoformat(span["end_time"][:-1])
            < utc_local_diff
        ):
            del span["end_time"]
        if (
            datetime.now() - datetime.fromisoformat(span["start_time"][:-1])
            < utc_local_diff
        ):
            del span["start_time"]

        self.found_string_spans.append(json.dumps(span))

    @staticmethod
    def flush():
        print("flushed the toilet")


def create_tracer_and_gh_settings(repo, actor, sha, run_id):
    tracing_attributes = {
        "build.repository": repo,
        "build.actor": actor,
    }
    resource = Resource(attributes={SERVICE_NAME: "velo-action", **tracing_attributes})
    trace.set_tracer_provider(TracerProvider(resource=resource))

    span_list = SpanList()
    trace.get_tracer_provider().add_span_processor(
        BatchSpanProcessor(ConsoleSpanExporter(out=span_list))
    )

    tracer = trace.get_tracer(__name__)

    gh_settings = GithubSettings(
        workspace="velo-action",
        sha=sha,
        ref_name="main",
        server_url="https://github.com",
        repository=repo,
        actor=actor,
        api_url="https://api.github.com",
        run_id=run_id,
        workflow="ci",
    )
    return tracer, gh_settings, span_list


def verify_results(tracer, span_list, results_file):
    """
    Compare the dynamically created spans with the expected results from a results_file.

    The main difficulty is that some properties are dynamic and cannot be compared. The trick is to simply delete these
    in both the stored results and in the spans created by the test before doing the comparison.
    """
    assert tracer.span_processor.force_flush()

    with (Path(__file__).parent / results_file).open("r") as file:
        required_spans = json.load(file)

    # If no start or end time exists while the spans are created they default to 'now'.
    # The solution is to delete all dynmically created 'now' timestamps from the spans.
    # The end time of the "build and deploy" span is always dynamic so we can delete all timestamps similar to that one.
    dynamic_end_time = [s for s in required_spans if s["name"] == "build and deploy"]
    dynamic_end_time = datetime.fromisoformat(dynamic_end_time[0]["end_time"][:-1])
    for span in required_spans:
        # Delete all dynamic attributes
        del span["resource"]["build.repository"]
        del span["resource"]["build.actor"]
        del span["context"]["trace_id"]
        del span["context"]["span_id"]
        del span["parent_id"]

        # Delete timestamps if they are dynamic
        end = datetime.fromisoformat(span["end_time"][:-1])
        if dynamic_end_time - end < timedelta(seconds=1):
            del span["end_time"]
        start = datetime.fromisoformat(span["start_time"][:-1])
        if dynamic_end_time - start < timedelta(seconds=1):
            del span["start_time"]

        # See if the span indeed exists
        try:
            span_list.found_string_spans.remove(json.dumps(span))
        except ValueError:
            pytest.fail(f"Span {span} not found")

    # All spans should have been found so this list should be empty now
    assert not span_list.found_string_spans


@has_token
def test_trace_creation_from_gh_response():
    """
    This test checks that the trace generation code outputs good traces. It is basically end to end testing hitting the
    GH API retrieving data on a workflow and then using the opentelemtry library to manually fill in traces.

    The hardcoded data has been generated by running using `make run` and copying it into a file. The only thing
    required to get it to work is to set an INPUT_TOKEN in env.dev-vars.

    You can find yours by running `gh auth status -t` if you have the GH cli installed.
    """
    tracer, gh_settings, span_list = create_tracer_and_gh_settings(
        "kolonialno/velo-action",
        "andersliland",
        "cbb6bfc4abc38abb8f50c60a567daac359c641cf",
        "2486719199",  # I think only this one matters?
    )

    construct_github_action_trace(
        tracer,
        gh_token(),
        "",
        github_settings=gh_settings,
    )
    verify_results(tracer, span_list, "action_trace_output_live_gh_api.json")


class MockResponse:
    def __init__(self):
        with (Path(__file__).parent / "gh_api_stored_response.json").open("r") as file:
            self.response = json.load(file)

    def json(self):
        return self.response

    def raise_for_status(self):
        pass


@patch("velo_action.github.requests.get", return_value=MockResponse())
def test_trace_creation_from_hardcoded_response(mocked_decode: MagicMock):
    """
    This test is very similar to the one in test_tracing_gh_api.py. The only difference is that this test uses a
    hardcoded gh api response from a workflow that is running instead of done.
    """
    tracer, gh_settings, span_list = create_tracer_and_gh_settings(
        "kolonialno/alma",
        "frenor",
        "a88bf7ccb2505624925a9e74175d2c01ec681591",
        "2795217427",
    )

    construct_github_action_trace(
        tracer,
        "notoken",  # This works because the request is mocked by 'patch'
        "",
        github_settings=gh_settings,
    )

    verify_results(tracer, span_list, "action_trace_output_hardcoded_gh_api.json")
    mocked_decode.assert_called_once()


velo_action/tests/test_tracing_gh_api.py
import json
import os
import re
import subprocess
from pathlib import Path

import pytest
from opentelemetry import trace
from opentelemetry.sdk.resources import SERVICE_NAME, Resource  # type: ignore
from opentelemetry.sdk.trace import TracerProvider  # type: ignore
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter  # type: ignore

from velo_action.settings import GithubSettings
from velo_action.tracing_helpers import construct_github_action_trace


def gh_token():
    if token := os.environ.get("GITHUB_TOKEN"):
        return token
    try:
        result = subprocess.run(
            ["gh", "auth", "status", "-t"], capture_output=True, text=True, check=True
        )
        return re.search(r"Token: (.*)\n", result.stderr).group(1)
    except Exception:  # pylint: disable=broad-except
        return None


has_token = pytest.mark.skipif(not bool(gh_token()), reason="No Github token found")


class SpanList:
    def __init__(self):
        self.found_string_spans = []

    def write(self, span):
        span = json.loads(span)

        del span["context"]["trace_id"]
        del span["context"]["span_id"]
        del span["parent_id"]

        if span["name"] == "build and deploy":
            del span["end_time"]  # Disregard because of dynamic end_time

        self.found_string_spans.append(json.dumps(span))

    @staticmethod
    def flush():
        print("flushed the toilet")


@has_token
def test_trace_creation():
    """
    This test checks that the trace generation code outputs good traces. It is basically end to end testing hitting the
    GH API retrieving data on a workflow and then using the opentelemtry library to manually fill in traces.

    The hardcoded data has been generated by running using `make run` and copying it into a file. The only thing
    required to get it to work is to set an INPUT_TOKEN in env.dev-vars.

    You can find yours by running `gh auth status -t` if you have the GH cli installed.
    """

    workspace = "velo-action"
    repo = f"kolonialno/{workspace}"
    actor = "andersliland"

    tracing_attributes = {
        "build.repository": repo,
        "build.actor": actor,
    }
    resource = Resource(attributes={SERVICE_NAME: "velo-action", **tracing_attributes})
    trace.set_tracer_provider(TracerProvider(resource=resource))

    span_list = SpanList()
    trace.get_tracer_provider().add_span_processor(
        BatchSpanProcessor(ConsoleSpanExporter(out=span_list))
    )

    tracer = trace.get_tracer(__name__)

    gh_settings = GithubSettings(
        workspace=workspace,
        sha="cbb6bfc4abc38abb8f50c60a567daac359c641cf",
        ref_name="main",
        server_url="https://github.com",
        repository=repo,
        actor=actor,
        api_url="https://api.github.com",
        run_id="2486719199",
        workflow="ci",
    )

    construct_github_action_trace(
        tracer,
        gh_token(),
        "",
        github_settings=gh_settings,
    )

    assert tracer.span_processor.force_flush()

    with (Path(__file__).parent / "action_trace_output.json").open("r") as file:
        required_spans = json.load(file)

    for span in required_spans:
        del span["context"]["trace_id"]
        del span["context"]["span_id"]
        del span["parent_id"]

        if span["name"] == "build and deploy":
            del span["end_time"]  # Disregard because of dynamic end_time
        span_list.found_string_spans.remove(json.dumps(span))

    assert not span_list.found_string_spans


velo_action/main.py
import os
import sys
from pathlib import Path

import pydantic
from loguru import logger

from velo_action import gcp, github
from velo_action.octopus.client import OctopusClient
from velo_action.octopus.deployment import Deployment
from velo_action.octopus.release import Release
from velo_action.settings import (
    VELO_TRACE_ID_NAME,
    ActionInputs,
    ActionOutputs,
    GithubSettings,
    resolve_workspace,
)
from velo_action.tracing_helpers import (
    construct_github_action_trace,
    init_tracer,
    print_trace_link,
    stringify_span,
)
from velo_action.utils import read_velo_settings

BASE_DIR = Path(__file__).resolve().parent.parent

VELO_DEPLOY_FOLDER_NAME = ".deploy"
LOG_FORMAT = "{time:YYYY-MM-DD HH:mm:ss} {message}"


def action(  # pylint: disable=too-many-branches,too-many-locals,too-many-statements
    args: ActionInputs,
    github_settings: GithubSettings,
) -> ActionOutputs:
    """Velo-action has two seperate usages: to create and deplot a release, or just generate verison.

    When just generating version it will be run without any inputs.
    Meaning no 'service_account_key'.
    This should not produce an error when initialising the tracing.
    """
    init_trace = False

    local_debug_mode = pydantic.parse_obj_as(
        bool, os.getenv("LOCAL_DEBUG_MODE", "False")
    )
    if args.service_account_key or local_debug_mode:
        # Do not init tracer when action is running without a
        # service_account_key.
        # This is supported behavior when only generating the verison.
        try:
            tracer = init_tracer(args, github_settings)
            span = construct_github_action_trace(
                tracer,
                args.token,
                args.preceding_run_ids,
                github_settings=github_settings,
            )
            trace_id = stringify_span(span)
            init_trace = True
        except Exception as error:  # pylint: disable=broad-except
            trace_id = None
            logger.warning(f"Starting trace failed: {error}", exc_info=error)

    if args.create_release or args.deploy_to_environments:
        deploy_folder = Path.joinpath(Path(args.workspace), VELO_DEPLOY_FOLDER_NAME)  # type: ignore
        if not deploy_folder.is_dir():
            sys.exit(
                f"Did not find a '{VELO_DEPLOY_FOLDER_NAME}' folder in '{args.workspace}'."
            )

        os.chdir(args.workspace)  # type: ignore

        gcloud = gcp.GCP(
            project=args.velo_project, service_account_key=args.service_account_key
        )
        octopus_server = gcloud.lookup_data(
            args.octopus_server_secret, args.velo_project
        )
        octopus_api_key = gcloud.lookup_data(
            args.octopus_api_key_secret, args.velo_project
        )
        velo_artifact_bucket = gcloud.lookup_data(
            args.velo_artifact_bucket_secret, args.velo_project
        )
        octo = OctopusClient(
            server=octopus_server,
            api_key=octopus_api_key,
            auth_token=args.service_account_key,
        )

    if args.create_release:
        release = Release(client=octo)

        velo_settings = read_velo_settings(deploy_folder)

        release_exists = release.exists(
            project_name=velo_settings.project, version=args.version, client=octo
        )

        if release_exists:
            logger.info(
                f"Release '{args.version}' already exists at "
                f"'{release.client.baseurl}/app#/Spaces-1/projects/"
                f"{velo_settings.project}/deployments/releases/{args.version}'. "
                "If you want to recreate this release, please delete it first in Octopus Deploy."
                "Project -> Releases -> <Select Release> -> : menu in top right corner -> Delete. "
            )
        else:
            files = gcloud.upload_from_directory(
                path=deploy_folder,
                dest_bucket_name=velo_artifact_bucket,
                dest_blob_name=f"{velo_settings.project}/{args.version}",
            )

            logger.info(
                f"Uploaded {len(files)} release files to "
                "'https://console.cloud.google.com/storage/browser/"
                f"{velo_artifact_bucket}/{velo_settings.project}/{args.version}'"
            )

            logger.info(
                f"Creating a release in Octopus Deploy for project '{velo_settings.project}' with version '{args.version}'"
            )
            release.create(
                project_name=velo_settings.project,
                project_version=args.version,
                github_settings=github_settings,
            )
            logger.info(
                f"See {release.client.baseurl}/app#/Spaces-1/projects/{velo_settings.project}/deployments/releases/{args.version}"
            )

    if args.deploy_to_environments:
        logger.info(f"Deploy to environments: {args.deploy_to_environments}")
        deploy_vars = {}
        if trace_id:
            deploy_vars[VELO_TRACE_ID_NAME] = trace_id

        tenants = args.tenants or [None]  # type: ignore

        for env in args.deploy_to_environments:
            for ten in tenants:
                log = f"Deploying project '{velo_settings.project}' version '{args.version}' to '{env}' "
                if ten:
                    log += f"for tenant '{ten}'"
                logger.info(log)

                deploy = Deployment(
                    project_name=velo_settings.project,
                    version=args.version,
                    client=octo,
                )

                deploy.create(
                    env_name=env,
                    tenant=ten,
                    wait_seconds=args.wait_for_success_seconds,
                    variables=deploy_vars,
                )

    if init_trace and (args.deploy_to_environments or args.create_release):
        print_trace_link(span)

    output = ActionOutputs(version=args.version)
    # Set outputs in environment to be used by other
    # steps in the Github Action Workflows
    logger.info("Github actions outputs:")
    github.actions_output("version", args.version)

    return output


if __name__ == "__main__":
    try:
        gh = GithubSettings()
        s = ActionInputs()

        s.workspace = resolve_workspace(s, gh)

    except pydantic.ValidationError as err:
        # Logger is not instantiated yet
        print(err)
        sys.exit(1)

    logger.add(sys.stdout, level=s.log_level, format=LOG_FORMAT)
    action(s, gh)


velo_action/utils.py
import os
import time
import json
import jwt
from pathlib import Path
from pathlib import Path
from typing import List, Optional

from semantic_version import SimpleSpec, Version

from velo_action.settings import (
    APP_SPEC_FIELD_PROJECT,
    APP_SPEC_FILENAMES,
    VeloSettings,
)


def resolve_app_spec_filename(deploy_folder: Path) -> Path:
    for filename in APP_SPEC_FILENAMES:
        filepath = Path.joinpath(deploy_folder, filename)
        if filepath.is_file():
            return filepath
    raise FileNotFoundError(
        f"Did not find an app.yml or app.yaml file in '{deploy_folder}'"
    )


def read_file(file: Path):
    if not os.path.exists(file):
        raise FileNotFoundError(f"{file} does not exist")

    with open(file, "r", encoding="utf-8") as stream:
        return stream.read()


def read_velo_settings(deploy_folder: Path) -> VeloSettings:
    """Parse the AppSpec (app.yml)"""
    filepath = resolve_app_spec_filename(deploy_folder)

    try:
        project = read_field_from_app_spec(APP_SPEC_FIELD_PROJECT, filepath)
    except ValueError as error:
        raise SystemExit(
            "'project' field is required in the AppSpec (app.yml). "
            "See https://centro.prod.nube.tech/docs/default/component/velo/app-spec/ for instructions."
        ) from error

    return VeloSettings(project=project)


def read_field_from_app_spec(field: str, filename: Path) -> str:
    """Read the project field from the app.yml.

    Cannot assume the app.yml is rendered,
    hence we cannot read the files as YAML.
    """
    with open(filename, encoding="utf-8") as file:
        lines = file.readlines()
        for line in lines:
            if line.startswith(f"{field}:"):
                return line.split(":")[1].strip().strip('"').strip("'")
    raise ValueError(f"Could not find '{field}' in {filename}")


def find_matching_version(
    versions: List[str], version_to_match: SimpleSpec
) -> Optional[Version]:
    """
    Finds the highest matching version in a list of versions.
    using the python semantic_version package.
    """
    versions = [Version.coerce(v) for v in versions]
    matching_version = version_to_match.select(versions)
    return matching_version


def create_self_signed_jwt(jwt_content, url):

    gsa = json.loads(jwt_content)

    iat = time.time()
    exp = iat + 3600
    payload = {
        "email": gsa["client_email"],
        "iss": gsa["client_email"],
        "sub": gsa["client_email"],
        "aud": url,
        "iat": iat,
        "exp": exp,
    }
    additional_headers = {"kid": gsa["private_key_id"]}

    signed_jwt = jwt.encode(
        payload,
        gsa["private_key"],
        headers=additional_headers,
        algorithm="RS256",
    )

    return signed_jwt


def is_valid_gsa_json(gsa_token_json):
    try:
        json.loads(gsa_token_json)
    except ValueError as e:
        return False
    return True


velo_action/github.py
import os

import requests
from loguru import logger

from velo_action.settings import GithubSettings


def actions_output(key, value):
    """Output variables such that they can be used in GitHub Actions."""
    os.system(f'echo "::set-output name={key}::{value}"')
    logger.info(f"{key}: {value}")


def request_github_workflow_data(
    token: str, preceding_run_ids: str, github_settings: GithubSettings
):
    github_headers = {"authorization": f"Bearer {token}"}

    base_url = (
        f"{github_settings.api_url}/repos/{github_settings.repository}/actions/runs"
    )
    current_workflow_url = f"{base_url}/{github_settings.run_id}/jobs"

    req = requests.get(current_workflow_url, headers=github_headers)
    req.raise_for_status()
    total_action_dict = {github_settings.workflow: req.json()}

    if preceding_run_ids:
        for preceding_run_id in preceding_run_ids.split(","):
            workflow_run_url = f"{base_url}/{preceding_run_id}"

            req = requests.get(workflow_run_url, headers=github_headers)
            req.raise_for_status()
            preceding_wf_name = req.json()["name"]

            req = requests.get(f"{workflow_run_url}/jobs", headers=github_headers)
            req.raise_for_status()
            total_action_dict[preceding_wf_name] = req.json()

    for wf_jobs in total_action_dict.values():
        for job in wf_jobs["jobs"]:
            for step in job["steps"]:
                try:
                    if step["started_at"] > step["completed_at"]:
                        step["completed_at"] = step["started_at"]
                except TypeError:
                    logger.debug(f"{step['name']} has no completed_at")
    return total_action_dict


def request_commit_info(
    token: str, commit_sha: str, github_settings: GithubSettings
) -> dict:
    if not token:
        return {"commit": {"message": "Unknown"}}

    github_headers = {"authorization": f"Bearer {token}"}
    req = requests.get(
        f"{github_settings.api_url}/repos/{github_settings.repository}/commits/{commit_sha}",
        headers=github_headers,
    )
    req.raise_for_status()
    return req.json()


velo_action/tracing_helpers.py
import base64
import datetime as dt
import json
import os
import time
from typing import Any

import jwt
import pydantic
from loguru import logger
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import (  # type: ignore
    OTLPSpanExporter,
)
from opentelemetry.sdk.resources import SERVICE_NAME, Resource  # type: ignore
from opentelemetry.sdk.trace import TracerProvider  # type: ignore
from opentelemetry.sdk.trace.export import (  # type: ignore
    BatchSpanProcessor,
    ConsoleSpanExporter,
)
from opentelemetry.trace import set_span_in_context
from opentelemetry.trace.status import Status, StatusCode

from velo_action.github import request_github_workflow_data
from velo_action.settings import GRAFANA_URL, ActionInputs, GithubSettings


def init_tracer(
    args: ActionInputs,
    github_settings: GithubSettings,
) -> TracerProvider:
    base_url = f"{github_settings.server_url}/{github_settings.repository}/actions/runs"
    workflow_url = f"{base_url}/{github_settings.run_id}"

    tracing_attributes = {
        "build.repository": github_settings.repository,
        "build.actor": github_settings.actor,
        "build.sha": github_settings.sha,
        "build.workflow_url": workflow_url,
    }
    resource = Resource(attributes={SERVICE_NAME: "velo-action", **tracing_attributes})
    trace.set_tracer_provider(TracerProvider(resource=resource))

    # Defaults to local_debug_mode = False
    if pydantic.parse_obj_as(bool, os.getenv("LOCAL_DEBUG_MODE", "False")):
        exporters = [
            ConsoleSpanExporter(),
            OTLPSpanExporter(
                endpoint="http://localhost:4318/v1/traces",
            ),
        ]
    else:
        jwt_content = json.loads(base64.b64decode(args.service_account_key).decode("ascii"))  # type: ignore
        iat = time.time() - 10
        exp = iat + 3600
        payload = {
            "iss": jwt_content["client_email"],
            "sub": jwt_content["client_email"],
            "aud": "some-cool-internally-nube-app-endpoint",
            "email": jwt_content["client_email"],
            "iat": iat,
            "exp": exp,
        }
        additional_headers = {"kid": jwt_content["private_key_id"]}
        signed_jwt = jwt.encode(
            payload,
            jwt_content["private_key"],
            headers=additional_headers,
            algorithm="RS256",
        )
        headers = {"Authorization": f"Bearer {signed_jwt}"}

        exporters = [
            OTLPSpanExporter(
                endpoint="https://traces-http.infra.nube.tech/v1/traces",
                headers=headers,
            ),
            OTLPSpanExporter(
                endpoint="https://otel.infra.nube.tech/v1/traces",
                headers=headers,
            ),
        ]

    for exporter in exporters:
        trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(exporter))

    return trace.get_tracer(__name__)


def print_trace_link(span: Any) -> None:
    trace_host = GRAFANA_URL
    # Use this locally together with docker-compose in the velo-tracing directory
    # trace_host = "http://localhost:3000"
    logger.info(
        f"See trace: {trace_host}/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Tem"
        f"po%22,%7B%22queryType%22:%22traceId%22,%22query%22:%22{span.context.trace_id:x}%22%7D%5D"
    )


def convert_time(input_time):
    return (
        round(dt.datetime.fromisoformat(input_time.replace("Z", "+00:00")).timestamp())
        * 1000000000  # required factor for otel span timing nanoseconds I guess?
    )


def trace_jobs(wf_jobs):
    start_times = []
    end_times = []
    job_spans = []

    for job in wf_jobs["jobs"]:
        if job["status"] == "queued":
            continue  # Do not trace jobs that are in the future

        new_job_span = make_empty_span_dict(job["name"])
        if job["started_at"]:
            span_start = convert_time(job["started_at"])
            new_job_span["start"] = span_start
            start_times.append(span_start)

        if job["completed_at"]:
            span_end = convert_time(job["completed_at"])
            new_job_span["end"] = span_end
            end_times.append(span_end)

        for step in job["steps"]:
            new_step_span = {
                "name": step["name"],
                "status": step["status"],
                "conclusion": step["conclusion"],
                "start": 0,
                "end": 0,
                "sub_spans": [],
            }
            if step["started_at"]:
                new_step_span["start"] = convert_time(step["started_at"])
            if step["completed_at"]:
                new_step_span["end"] = convert_time(step["completed_at"])
            new_job_span["sub_spans"].append(new_step_span)
        job_spans.append(new_job_span)
    end_time = max(end_times) if end_times else 0
    return min(start_times), end_time, job_spans


def recurse_add_spans(tracer, parent_span, sub_span_dict):
    # 0 is falsy so this will take now instead of 0 if the time isn't set
    span = tracer.start_span(
        sub_span_dict["name"],
        start_time=sub_span_dict["start"] or None,
        context=set_span_in_context(parent_span),
    )
    sub_span_dict["span"] = span
    for span_dict in sub_span_dict["sub_spans"]:
        recurse_add_spans(tracer, span, span_dict)

    if sub_span_dict.get("status") == "failure":
        span.set_status(Status(StatusCode.ERROR, sub_span_dict["conclusion"]))
    span.end(sub_span_dict["end"] or None)


def make_empty_span_dict(name, start=0, end=0, sub_spans=None, span=None):
    return {
        "name": name,
        "start": start,
        "end": end,
        "sub_spans": sub_spans if sub_spans is not None else [],
        "span": span,
    }


def stringify_span(span):
    return f"{span.context.trace_id:x}:{span.context.span_id:x}:0:{span.context.trace_flags:x}"


def construct_github_action_trace(
    tracer, token: str, preceding_run_ids: str, github_settings: GithubSettings
) -> Any:

    total_action_dict = request_github_workflow_data(
        token=token,
        preceding_run_ids=preceding_run_ids,
        github_settings=github_settings,
    )

    wf_start_times = []
    span_dict = make_empty_span_dict("build and deploy")

    for wf_name, wf_jobs in total_action_dict.items():
        wf_start_time, wf_end_time, job_spans = trace_jobs(wf_jobs)
        wf_start_times.append(wf_start_time)
        span_dict["sub_spans"].append(
            make_empty_span_dict(wf_name, wf_start_time, wf_end_time, job_spans)
        )
    span_dict["start"] = min(wf_start_times) - 1

    # Convert dict into actual opentelemetry spans!
    span = tracer.start_span(span_dict["name"], start_time=span_dict["start"])
    span_dict["span"] = span
    for wf_span_dict in span_dict["sub_spans"]:
        recurse_add_spans(tracer, span, wf_span_dict)
        if wf_span_dict["name"] == github_settings.workflow:
            logger.debug(f"Current wf_span: {stringify_span(wf_span_dict['span'])}")
    span.end(span_dict["end"] or None)
    return span
